pkg/traceql/ast_execute.go

```diff
@@ -428,10 +448,40 @@ func (o *BinaryOperation) execute(span Span) (Static, error) {
 		lhsB, _ := lhs.Bool()
 		rhsB, _ := rhs.Bool()
 
+		if recording {
+			if done := o.b.Sampled(); done {
+				if o.b.OptimalBranch() == 1 {
+					// RHS is the optimal starting branch,
+					// so swap the elements now.
+					o.LHS, o.RHS = o.RHS, o.LHS
```

Comment:
This is the final magic statement that flips the expression.

Reply:
Agree, this order isn't perfect since it will only actually do numSamples-1 samples.   The reason I didn't is that I was assuming that defering it would be too slow, and so would any changes to the flow:  we would have to calculate the return value without returning from the switch statement, then sample, then return.  But I can try it.

Reply:
Thanks for pushing on this. Ended up with a double switch statement which I thought would be slower but it is faster because it ensures all recording logic is under one if statement.  It gave a few % regression back.

Here are the main two regressions now:
```
BackendBlockTraceQL/traceOrMatch                     530ms ± 2%     545ms ± 0%  +2.95%  (p=0.004 n=6+5)
BackendBlockTraceQL/traceOrNoMatch                   533ms ± 1%     552ms ± 1%  +3.60%  (p=0.002 n=6+6)
```